src/components/implementation/no_interface/cFE_booter/osfilesys.c

```diff
@@ -1,280 +1,886 @@
+
+    if (strcmp(token, cur->name) != 0) return OS_FS_ERR_PATH_INVALID;
+    //loop terminates when it finds a place to put o or determines path is invalid
+
+    while (1) {
+        if (token == NULL) {
+            return OS_FS_ERR_PATH_INVALID;
+        }
+        assert(cur->name);
+
+        //if there is no child, then insert as child
+        if (!cur->child) {
+            // if the next part of the path is not o->name or there is a part after it, bad path
+            token = strtok(NULL, delim);
+            if (strcmp(token, o->name) != 0 || strtok(NULL, delim) != NULL) {
+                return OS_FS_ERR_PATH_INVALID;
+            }
+
+            o->parent = cur;
+            cur->child = o;
+            o->next = NULL;
+            o->prev = NULL;
+            return OS_FS_SUCCESS;
+        }
+        cur = cur->child;
+        token = strtok(NULL, delim);
+
+        //precondition: cur is the first in a non-empty list of children
+        //postcondition: cur is an ancestor of o or o has been inserted in list
+        // while cur is not ancestor of o
+        while (strcmp(token, cur->name) != 0) {
+            if (cur->next == NULL) {
+                // if the next part of the path is o->name or there is a part after it, bad path
+                if (strcmp(token, o->name) != 0 || strtok(NULL, delim) != NULL) {
+                    return OS_FS_ERR_PATH_INVALID;
+                }
+                // insert o as the last child in a linked list of children
+                cur->next = o;
+                o->prev = cur;
+                o->parent = cur->parent;
+                o->next = NULL;
+                return OS_FS_SUCCESS;
+            }
+            cur = cur->next;
+        }
+    }
+    PANIC("Unreachable Statement");
+    return 0;
 }
 
-/*
- * Copies information from a tar file header to a fsobj
- */
-uint32 tar_cphdr(uint32 tar_offset, struct fsobj *file)
+int32 file_open(char *path, permission_t permission)
+{
+    assert(openfs);
+    if (!openfs->root) return OS_FS_ERR_PATH_TOO_LONG;
+    if (!path) return OS_FS_ERR_INVALID_POINTER;
+
+    // find the file
+    struct fsobj *file = file_find(path);
+    if (!file) return OS_FS_ERR_PATH_INVALID;
+    if (file->type != FSOBJ_FILE) return OS_FS_ERROR;
+
+    // get a new fd
+    int32 FD = newfd_get(file->ino);
+    if (FD == OS_FS_ERROR) {
+        return OS_FS_ERR_NO_FREE_FDS;
+    }
+
+    struct fd *filedes = &fd_tbl[FD];
+
+    filedes->access                         = permission;
+    filedes->file                           = file;
+    filedes->position.file_pos.open_part    = file->file_part;
+    filedes->position.file_pos.file_offset  = 0;
+    filedes->position.file_pos.part_offset  = 0;
+    file->refcnt++;
+    return FD;
+}
+
+int32 file_close(int32 FD)
 {
-    //TODO: store filenames without paths
-    assert(tar_offset < size);
-    assert(file->filedes > 0);
+    int32 ret = chk_fd(FD);
+    if (ret != OS_FS_SUCCESS) return OS_FS_ERROR;
```

Comment:
chk_fd returns invalid fd code, but the api specifies that close will only return success or error.